### PR TITLE
perf(ssz): setup buffers for memory re-use

### PR DESCRIPTION
**Pull request by calbera**
Next steps:
- Deprecate Tree from pkg/merkle in favor of pkg/ssz/merkle
- Create and store Hasher object at high level (state processor) for re-use
  - passed in via `.HashTreeRoot(/* hasher */)`
- Minor: add U64 generic on Hasher object

## Summary by CodeRabbit

- **New Features**
  - Introduced reusable and single-use buffers for Merkle tree hashing, optimizing memory usage.
  - Added a `Hasher` type for constructing Merkle tree roots, enhancing modularity and performance.

- **Bug Fixes**
  - Updated Merkle tree root generation to handle edge cases like empty leaf sets more robustly.

- **Performance Improvements**
  - Implemented benchmarks for reusable and single-use buffer functionalities in Merkle hashing to ensure optimal performance.

- **Testing**
  - Enhanced test coverage for Merkle tree root calculations and buffer functionalities with new test cases and benchmarks.

### CODE REVIEW

File: mod/primitives/pkg/merkle/buffer_test.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			}
+
+			result[0] = [32]byte{}
+			result[0][i] = byte(i)
+		}
+	}
+}
+
+// Benchmark for the Get method
+//
```
Comment (itsdevbear) [inline mod/primitives/pkg/merkle/buffer_test.go:74]: ur gonna wanna benchmark the BuildRoots before and after the buf

File: mod/primitives/pkg/ssz/merkleize.go [hunk 2]
```diff
@@ -247,8 +247,10 @@ func Merkleize[U64T U64[U64T], RootT ~[32]byte](
 		return effectiveChunks[0], nil
 	}
 
-	return merkle.NewRootWithMaxLeaves(
+	// TODO: reuse the same hasher for reduced memory allocations.
+	hasher := merkle.NewHasher[RootT]()
+	return hasher.NewRootWithMaxLeaves(
 		effectiveChunks,
-		effectiveLimit,
+		uint64(effectiveLimit),
```
Comment (calbera) [inline mod/primitives/pkg/ssz/merkleize.go:254]: actual change here

File: mod/primitives/pkg/merkle/tree.go [hunk 3]
```diff
@@ -82,9 +82,10 @@ func NewTreeFromLeavesWithDepth[RootT ~[32]byte](
 
 	// Preallocate layers based on depth
 	// TODO: This should be done virtually....
+	buffer := NewSingleuseBuffer[RootT]()
 	for i := uint8(1); i <= depth; i++ {
 		layerSize := (len(leaves) + (1 << i) - 1) >> i
-		layers[i] = make([]RootT, layerSize)
+		layers[i] = buffer.Get(layerSize)
```
Comment (calbera) [inline mod/primitives/pkg/merkle/tree.go:88]: actual change here

File: mod/primitives/pkg/merkle/hasher_test.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, err := hasher.NewRootWithMaxLeaves(leaves, uint64(len(leaves)))
+		if err != nil {
+			b.Fatalf("Expected no error, got %v", err)
+		}
+	}
+}
```
Comment (calbera) [inline mod/primitives/pkg/merkle/hasher_test.go:135]: important benchmark

**Pull request merged**